File: moseoh/_10주차/숫자_변환하기/java/Solution.java
package _10주차.숫자_변환하기.java;

import java.util.Arrays;

class Solution {
    public static void main(String[] args) {
        int x1 = 10;
        int y1 = 40;
        int n1 = 5;
        int answer1 = 2;
        int result1 = new Solution().solution(x1, y1, n1);
        PRINT_RESULT(1, result1, answer1);

        int x2 = 10;
        int y2 = 40;
        int n2 = 30;
        int answer2 = 1;
        int result2 = new Solution().solution(x2, y2, n2);
        PRINT_RESULT(2, result2, answer2);

        int x3 = 2;
        int y3 = 5;
        int n3 = 4;
        int answer3 = -1;
        int result3 = new Solution().solution(x3, y3, n3);
        PRINT_RESULT(3, result3, answer3);
    }

    public static void PRINT_RESULT(int index, int result, int answer) {
        boolean correct = result == answer;
        StringBuilder sb = new StringBuilder();
        sb.append("\n\n테스트 케이스 ").append(index).append(": ");
        sb.append(correct ? "정답" : "오답").append("\n");
        sb.append("\t- 실행 결과: \t").append(result).append("\n");
        sb.append("\t- 기댓값: \t").append(answer).append("\n");
        if (correct) System.out.println(sb);
        else throw new RuntimeException(sb.toString());
    }

    public int solution(int x, int y, int n) {
        int[] dp = new int[y + 1];
        Arrays.fill(dp, -1);
        dp[x] = 0;
        for (int i = x + 1; i <= y; i++) {
            int min = Integer.MAX_VALUE;
            if (i % 3 == 0 && dp[i / 3] != -1) min = Math.min(dp[i / 3] + 1, min);
            if (i % 2 == 0 && dp[i / 2] != -1) min = Math.min(dp[i / 2] + 1, min);
            if (i - n >= x && dp[i - n] != -1) min = Math.min(dp[i - n] + 1, min);
            if (min != Integer.MAX_VALUE) dp[i] = min;
        }
        return dp[y];
    }
}

File: moseoh/_10주차/숫자_변환하기/kotlin/Solution.kt
package _10주차.숫자_변환하기.kotlin


fun main() {
    fun printResult(index: Int, result: Int, answer: Int) {
        val correct = result == answer
        val sb = StringBuilder()
        sb.append("\n\n테스트 케이스 ").append(index).append(": ")
        sb.append(if (correct) "정답" else "오답").append("\n")
        sb.append("\t- 실행 결과: \t").append(result).append("\n")
        sb.append("\t- 기댓값: \t").append(answer).append("\n")
        if (correct) println(sb) else throw RuntimeException(sb.toString())
    }

    val x1 = 10
    val y1 = 40
    val n1 = 5
    val answer1 = 2
    val result1 = Solution().solution(x1, y1, n1)
    printResult(1, result1, answer1)

    val x2 = 10
    val y2 = 40
    val n2 = 30
    val answer2 = 1
    val result2 = Solution().solution(x2, y2, n2)
    printResult(2, result2, answer2)

    val x3 = 2
    val y3 = 5
    val n3 = 4
    val answer3 = -1
    val result3 = Solution().solution(x3, y3, n3)
    printResult(3, result3, answer3)
}

class Solution {
    fun solution(x: Int, y: Int, n: Int): Int {
        return IntArray(y + 1) { -1 }.apply {
            this[x] = 0
            for (i in x + 1..y) {
                this[i] = listOfNotNull(
                    if (i % 3 == 0 && this[i / 3] != -1) this[i / 3] + 1 else null,
                    if (i % 2 == 0 && this[i / 2] != -1) this[i / 2] + 1 else null,
                    if (i - n >= x && this[i - n] != -1) this[i - n] + 1 else null
                ).minOrNull() ?: -1
            }
        }[y]
    }
}
